torch/linalg/__init__.py

```diff
@@ -4,6 +4,8 @@
 import torch
 from torch._C import _add_docstr, _linalg  # type: ignore[attr-defined]
 
+LinAlgError = torch.LinAlgError # type: ignore[attr-defined]
```

Comment:
You have two strategies here:
- Keep it both in torch. and torch.linalg. in that case, you can add LinAlgError to the __all__ list in torch/__init__.py to fix this lint error.
- Only have it here, in that case, you can change the code in Exceptions.cpp to name it `_LinAlgError` and change the line here to `LinAlgError = torch._C._LinAlgError`.

Either way, it is missing doc (you can use the `_add_docs` function to easily add doc to a c++ object from python).

Reply:
I implemented the `_LinAlgError` as you suggested. However when I add docs using `_add_docstr` using the following code:

``` python

_add_docstr(torch._C._LinAlgError, r"""
Error raised by torch.linalg function when the cause of error is a numerical inconsistency in the data.
For example, you can the torch.linalg.inv function will raise torch.linalg.LinAlgError when it finds that
a matrix is not invertible.

Example:
    >>> matrix = torch.eye(3, 3)
    >>> matrix[-1, -1] = 0
    >>> matrix
    tensor([[1., 0., 0.],
            [0., 1., 0.],
            [0., 0., 0.]])
    >>> torch.linalg.inv(matrix)
    Traceback (most recent call last):
    File "<stdin>", line 1, in <module>
    torch._C._LinAlgError: torch.linalg.inv: The diagonal element 3 is zero, the inversion
    could not be completed because the input matrix is singular.
""")
```

I get a malloc error such as follows:

```
python(59485,0x200e09600) malloc: *** error for object 0x7f8b648e9ba0: pointer being freed was not allocated
python(59485,0x200e09600) malloc: *** set a breakpoint in malloc_error_break to debug
[1]    59485 abort      pytest test/test_linalg.py -k test_cholesky_ex_non_pd
```

Any idea how this can be resolved?

Reply:
I tried running through a debugger and it turns out there is a problem with deallocation of the Python string that is associated as a doc string with this method.

Reply:
wow I've never seen this one before :o 
Maybe do the same with the `LinAlgError` object in `linalg/__init__.py`?

Reply:
I tried that too but still the same. Possibly the add_docstr has some issue when dealing with error classes? (I cannot see another example where it is used for adds docs for errors)

Reply:
OK so I investigated the Python C API source code and the way they set the doc string for exception is by setting a dict and setting the `__doc__` value. My guess is that there is some issue with setting `tp_doc` for the doc string for Exception classes. As a possible workaround I've input the doc string using `PyExceptionWithDoc`. Do let me know if this works. I must admit that it does look kind of ugly and a way around `add_docstr` would be ideal.